packages/flutter/lib/src/widgets/nested_scroll_view.dart

```diff
@@ -2068,3 +2130,406 @@ class RenderNestedScrollViewViewport extends RenderViewport {
     properties.add(DiagnosticsProperty<SliverOverlapAbsorberHandle>('handle', handle));
   }
 }
+
+/// State of [NestedScrollController] :
+///  * Initially, the state is [_ControllerState.idle]
+///  * The [Scrollable] of outer scroll view creates [_NestedScrollPosition] object by calling
+///    [NestedScrollController.createScrollPosition].
+///  * [NestedScrollController] delegates the work to outerScrollController.
+///  * The [Scrollable] of outer scroll view attatch the outer position by calling
+///    [NestedScrollController.attach].
+///  * The state is now [_ControllerState.outerPositionAttached].
+///  * The [Scrollable] of inner scroll view creates [_NestedScrollPosition] object by calling
+///    [NestedScrollController.createScrollPosition].
+///  * [NestedScrollController] delegates the work to innerScrollController.
+///  * The [Scrollable] of inner scroll view attatch the inner position by calling
+///    [NestedScrollController.attach].
+///  * The state is now [_ControllerState.innerPositionAttached]. At this state,
+///    the creation and attachment of scroll positions are complete.
+///
+/// The logic of detaching is similar. The state goes from [_ControllerState.innerPositionAttached]
+/// to [_ControllerState.idle].
+///
+/// Creation of [_NestedScrollPosition] is always sequential(from outer to inner), but attachment is not.
+/// When [NestedScrollController] restore its scroll position using data saved by [RestorationManager],
+/// it may attach the scroll position in reverse order. [_NestedScrollPosition] distinguish
+/// the scroll position by [_NestedType]. Since the creation of [_NestedScrollPosition] is sequential,
+/// the [_NestedType] assigned to it is always accurate.
+enum _ControllerState {
+  idle,
+  outerPositionAttached,
+  innerPositionAttached,
+}
+
+enum _NestedType {
+  outer,
+  inner,
+}
+
+/// A controller for [NestedScrollView]
+///
+/// A NestedScrollController lets you manipulate the whole NestedScrollView. It holds two [ScrollController]
+/// which manipulate the header and body respectively.
+class NestedScrollController extends ScrollController with _CoordinatorMixin {
```

Comment:
It is confusing to have a NestedScrollController and a _NestedScrollController. I feel that maintaining this will be difficult to follow and reason about.